experimental/Pipeline/pipelie.mli

```diff
@@ -121,8 +128,77 @@ module Automation : sig
     | FromTopToCurrent
     | Eval
 
-  type t = {
+  type process_job = {
       current  : Sexp.t;
       newForms : (stage * Sexp.t) list
     }
+
+  type t =
+    | ProcessJob of process_job
+    | UpdateJob  of { new_context : Sexp.t Context.t; process : process_job }
+
+  type job = t
+
+  (** [pass_arguments] is a pipeline processing function, namely we wrap
+   * the current expression and the context into a single function *)
+  type pass_argument = {
+      current : Sexp.t;
+      context : Sexp.t Context.t
+    }
+
+  (* This represents the monad type that we program against *)
+  module type MonadEff = sig
+    type 'a t
+    val hasTrace : unit -> Trace.t
+    val hasMeta  : unit -> Meta.t
+    val run      : 'a t -> 'a
+  end
```

Comment:
Something I'm curious about here:  `run` looks to me like it has the type of `extract`, which is part of the signature of a `Comonad`, rather than a `Monad`.  `Reader` in particular has a `run` with that signature, but it _is_ a `Comonad` as well as a `Monad`.  So is this really more like `ComonadEff`?  Or is it also required to be a `Monad`?  Or am I just misunderstanding?  :)